crates/kubelet/src/module_store.rs

```diff
@@ -29,20 +63,85 @@ use std::sync::Arc;
 ///
 /// #[async_trait]
 /// impl ModuleStore for InMemoryStore {
-///     async fn get(&self, image_ref: &Reference) -> anyhow::Result<Vec<u8>> {
+///     async fn get_local(&self, image_ref: &Reference) -> anyhow::Result<Vec<u8>> {
 ///         match self.modules.get(image_ref) {
 ///             Some(bytes) => Ok(bytes.clone()),
 ///             None => todo!("Fetch the bytes from some sort of remore store (e.g., OCI Distribution)")
 ///         }
 ///     }
+///
+///     async fn is_present(&self, image_ref: &Reference) -> bool {
+///         self.modules.get(image_ref).is_some()
+///     }
+///
+///     async fn is_present_with_digest(&self, image_ref: &Reference, digest: String) -> bool {
+///         false
+///     }
+///
+///     async fn pull(&self, image_ref: &Reference) -> anyhow::Result<()> {
+///         Err(anyhow::anyhow!("InMemoryStore does not support registry pull"))
+///     }
+///
+///     async fn resolve_registry_digest(&self, image_ref: &Reference) -> anyhow::Result<String> {
+///         Err(anyhow::anyhow!("InMemoryStore does not support registry pull"))
+///     }
 /// }
 /// ```
 #[async_trait]
 pub trait ModuleStore {
     /// Get a module's data given its image `Reference`.
     ///
-    /// It is up to the implementation to establish caching and network fetching policies.
-    async fn get(&self, image_ref: &Reference) -> anyhow::Result<Vec<u8>>;
+    /// It is up to the implementation to establish caching policies.
+    /// However, the implementation must fail if the image is not present
+    /// locally.
+    async fn get_local(&self, image_ref: &Reference) -> anyhow::Result<Vec<u8>>;
+
+    /// Whether the specified module is already present in the store.
+    async fn is_present(&self, image_ref: &Reference) -> bool;
+
+    /// Whether the specified module is already present in the store with the specified digest.
+    async fn is_present_with_digest(&self, image_ref: &Reference, digest: String) -> bool;
```

Comment:
Just spitballing here: Because this store is k8s specific, would it be better to abstract away all these functions so we don't have to manage implementing and calling each of these separately? We can provide the default implementation here at least?

I dunno, maybe it wouldn't work.

Reply:
This is an excellent question and one of the things I was unsure about in the API design.  What I wanted to do was encode the mandated logic from https://kubernetes.io/docs/concepts/configuration/overview/#container-images so that other module stores didn't have to replicate that.  But that logic depends deeply on all the stuff about local presence and registry digests, so that means _implementers_ of ModuleStore need to surface that stuff for `get` to use, even though _consumers_ of ModuleStore shouldn't care.

So maybe the pattern is wrong here.  Maybe there should be a `ModuleStore` facade struct which exposes _only_ the `get` method (which requires a pull policy) - as a struct impl but not a trait - but the facade struct has `storage` and `fetcher` members which take care of the local cache and registry interaction side of things respectively.  `get` could use those members but they would be private and so not surfaced to users of the `ModuleStore`.  This shifts us away from the idea that `ModuleStore` is just a magic black box that you ask it for a module and lo! it dispenses one - it mandates some specifics about _how_ it works internally.

Though I guess we could still have a `ModuleStore` trait for the magic `get` dispenser interface, and a `TheOnlyModuleStoreYoullEverNeed` struct for the standard `get` implementation over arbitrary storers and fetchers.  That's probably nicest.

Bah, between this and that rogue Fisher moving all the files, I'm almost thinking I should start over.  The work and tests would come straight across but I'm not sure trying to change in place on top of a rebase/merge from hell is going to be worth it...!

Reply:
That idea sounds interesting for sure. Are you saying the `storage` and `fetcher` things would be Traits?

Reply:
Yes, like the `Client` trait we currently pass into `FileStore`.